### ISSUE TITLE
[RFC][serve] Serve application arguments

### ISSUE THREAD

**Issue by edoakes** [ray-project/ray#34542]
This RFC outlines a proposal to add support for arguments passed to Serve applications. This can be used for iterative development as well as parametrizing applications in production (e.g., use a different model path).

## Problem

Currently it’s difficult to parametrize Serve applications. Take the simple case where I have a wrapper that takes a model path, loads it, and calls it for inference:

```python
@serve.deployment
class ModelWrapper:
    def __init__(self, model_uri: str):
        self._model = load_model(model_uri)

    def __call__(self, *args):
        return self._model(*args)

app = ModelWrapper.bind("s3://model_uri_1")
```

If I want to pass a different model URI to this application (for iterative development or in production to update to a new version), there are currently three options:

- Modify the code itself to have the new path. Problem: I shouldn’t need to update my code to pass an argument!
- Define a `reconfigure` method, build a Serve config file `user_config`. Problem: This doesn’t fit into the `serve run` development workflow and it’s a lot of steps/code for something simple.
- Use environment variables and implement parsing/validation logic myself. This is an OK workaround but it's not very elegant and limits our ability to improve the UX (e.g., show the arguments in the dashboard).

## Proposal
In addition to supporting importing a built application object as we currently do, support builder functions that return a built app. These can take keyword arguments passed via the CLI or config file.

The motivation for a “builder” function is to easily extend to model composition: the flattened arguments can apply to any of the deployments in the application.

CLI usage:
```bash
serve run my_module:build_app model_uri="s3://model_uri_1"
# Easily iterate w/ different arguments.
serve run my_module:build_app model_uri="s3://model_uri_2"
```

Config usage:
```yaml
applications:
  my_app:
    import_path: my_module.build_app
    args:
      model_uri: s3://model_uri_1
```

Python API:
```python
# By default receive a dictionary containing passed args.
def build_app(args: Dict[str, str]) -> BoundDeployment:
    return ModelWrapper.bind(args["model_uri"])

# This can easily be extended to model composition:
def build_composed_app(args: Dict[str, str]) -> BoundDeployment:
    model = ModelWrapper.bind(args["model_uri"])
    return Preprocessor.bind(args["preprocessor_uri"], model)

# Typing support w/ pydantic:
from pydantic import BaseModel

class MyTypedArgs(BaseModel):
    model_uri: str
    model_num_cpus: int

def build_app_typed(args: MyTypedArgs) -> BoundDeployment:
    return Model.options(
        num_replicas=args.model_num_cpus
    ).bind(args.model_uri)
```

**Comment by edoakes** [comment]
Plan is to have this implemented for the next Ray release (2.5). Please leave any feedback about the use cases / API!

**Comment by edoakes** [comment]
@smit-kiri yes that will be supported 👍

**Comment by edoakes** [comment]
The proposed API has been added and will be released in Ray 2.5, please try it out!
